plugin.video.orftvthek/resources/lib/base.py

```diff
@@ -0,0 +1,242 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+import os
+import re
+
+import simplejson as json
+from kodi_six import xbmcplugin, xbmcgui, xbmcvfs
+from kodi_six.utils import py2_encode, py2_decode
+from . import Settings
+from .helpers import *
+
+
+def showDialog(title, description):
+    xbmcgui.Dialog().notification(title, description, xbmcaddon.Addon().getAddonInfo('icon'))
+
+
+def addDirectory(title, banner, backdrop, description, link, mode, pluginhandle):
+    parameters = {"link": link, "mode": mode}
+    u = build_kodi_url(parameters)
+    createListItem(title, banner, description, '', '', '', u, False, True, backdrop, pluginhandle, None)
+
+
+def generateAddonVideoUrl(videourl):
+    videourl = buildLink(videourl)
+    return "plugin://%s/?mode=play&link=%s" % (xbmcaddon.Addon().getAddonInfo('id'), videourl)
+
+
+def buildLink(link):
+    link = link.replace("https://apasfpd.apa.at", "https://apasfpd.sf.apa.at")
+    if link:
+        return "%s|User-Agent=%s" % (link, Settings.userAgent())
+    else:
+        return link
+
+
+def createPlayAllItem(name, pluginhandle, stream_info=False):
+    play_all_parameters = {"mode": "playlist"}
+    play_all_url = build_kodi_url(play_all_parameters)
+    play_all_item = xbmcgui.ListItem(name)
+    if stream_info:
+        description = stream_info['description']
+        play_all_item.setArt({'thumb': stream_info['teaser_image'], 'poster': stream_info['teaser_image']})
+    else:
+        description = ""
+    play_all_item.setInfo(type="Video", infoLabels={"Title": name, "Plot": description})
+    xbmcplugin.addDirectoryItem(pluginhandle, play_all_url, play_all_item, isFolder=False, totalItems=-1)
+
+
+def createListItem(title, banner, description, duration, date, channel, videourl, playable, folder, backdrop, pluginhandle, subtitles=None, blacklist=False, contextMenuItems=None):
+    contextMenuItems = contextMenuItems or []
+
+    liz = xbmcgui.ListItem(title)
```

Comment:
tip: adding `offscreen=True` whenever you create a `ListItem` will speed up your addon a bit

Reply:
nice, thx for the hint. i added this on all ListItem calls. couldnt find anything useful about this in the docs. what does it do?

Reply:
It avoids a GUI lock. Afaik this was not set as default due to compatibility reasons, long long ago.